src/librepgp/stream-write.c

```diff
@@ -0,0 +1,943 @@
+        dst->werr = RNP_ERROR_BAD_PARAMETERS;
+        return;
+    }
+
+    dst_write(param->writedst, buf, len);
+    dst->werr = RNP_SUCCESS;
+}
+
+void literal_dst_close(pgp_dest_t *dst, bool discard)
+{
+    pgp_dest_packet_param_t *param = dst->param;
+
+    if (!param) {
+        return;
+    }
+
+    close_streamed_packet(param, discard);
+    free(param);
+    dst->param = NULL;
+}
+
+pgp_errcode_t init_literal_dst(pgp_write_handler_t *handler, pgp_dest_t *dst, pgp_dest_t *writedst)
+{
+    pgp_dest_packet_param_t *param;
+    pgp_errcode_t            ret = RNP_SUCCESS;
+    int                      flen;
+    uint8_t                  buf[4];
+
+    if ((param = calloc(1, sizeof(*param))) == NULL) {
+        (void) fprintf(stderr, "init_literal_dst: allocation failed\n");
+        return RNP_ERROR_OUT_OF_MEMORY;
+    }
+
+    dst->param = param;
+    dst->writefunc = literal_dst_write;
+    dst->closefunc = literal_dst_close;
+    dst->type = PGP_STREAM_LITERAL;
+    param->partial = true;
+    param->indeterminate = false;
+    param->tag = PGP_PTAG_CT_LITDATA;
+
+    /* initializing partial length or indeterminate packet, writing header */
+    if (!init_streamed_packet(param, writedst)) {
+        (void) fprintf(stderr, "init_literal_dst: failed to init streamed packet\n");
+        ret = RNP_ERROR_BAD_PARAMETERS;
+        goto finish;
+    }
+    /* content type - forcing binary now */
+    buf[0] = (uint8_t)'b';
+    /* filename */
+    if (handler->ctx->filename) {
+        flen = strlen(handler->ctx->filename);
+        if (flen > 255) {
+            (void) fprintf(stderr, "init_literal_dst: filename too long, truncating\n");
+            flen = 255;
+        }
+    } else {
+        flen = 0;
+    }
+    buf[1] = (uint8_t)flen;
+    dst_write(param->writedst, buf, 2);
+    if (flen > 0) {
+        dst_write(param->writedst, handler->ctx->filename, flen);
+    }
+    /* timestamp */
+    buf[0] = (uint8_t)(handler->ctx->filemtime >> 24);
+    buf[1] = (uint8_t)(handler->ctx->filemtime >> 16);
+    buf[2] = (uint8_t)(handler->ctx->filemtime >> 8);
+    buf[3] = (uint8_t)(handler->ctx->filemtime);
+    dst_write(param->writedst, buf, 4);
+    
+    finish:
+    if (ret != RNP_SUCCESS) {
+        literal_dst_close(dst, true);
+    }
+
+    return ret;
+}
+
+pgp_errcode_t init_armored_dst(pgp_write_handler_t *handler, pgp_dest_t *dst, pgp_dest_t *writedst)
+{
+    return RNP_ERROR_NOT_IMPLEMENTED;
+}
+
+pgp_errcode_t rnp_encrypt_src(pgp_write_handler_t *handler, pgp_source_t *src, pgp_dest_t *dst)
+{
+    /* stack of the streams would be as following:
+       [armoring stream] - if armoring is enabled
+       encrypting stream, partial writing stream
+       [compressing stream, partial writing stream] - if compression is enabled
+       literal data stream, partial writing stream
+    */
+    uint8_t       readbuf[PGP_INPUT_CACHE_SIZE];
+    ssize_t       read;
+    pgp_dest_t    dests[4];
+    int           destc = 0;
```

Comment:
The `dests`+`destc` approach feels flimsy, it might be better to do something similar to the current `pgp_writer_push` approach..

Reply:
Internally dests have pointers to the next one. This array is to have access from the calling function to the whole list of dests. For instance, if some of them was failed to initialize and didn't put pointer to the correct place then we will not be able to dealloc things.